### ISSUE TITLE
Impersonation crashes due to lack of data validation

### ISSUE THREAD

**Issue by s-huk** [metabase/metabase#47296]
### Describe the bug

We experience this bug when using Impersonation in Metabase Enterprise (see below how to reproduce more exactly):
```
class clojure.lang.PersistentVector cannot be cast to class java.lang.CharSequence (clojure.lang.PersistentVector is in unnamed module of loader 'app'; java.lang.CharSequence is in module java.base of loader 'bootstrap')
...
metabase_enterprise.advanced_permissions.driver.impersonation$connection_impersonation_role.invokeStatic(impersonation.clj:73)
```
This means that [this line of Metabase Enterprise code crashes](https://github.com/metabase/metabase/blob/6bf7bfca9aa8a71ba8e155a2c90713d0d9999951/enterprise/backend/src/metabase_enterprise/advanced_permissions/driver/impersonation.clj#L73).

**I delivered a solution proposal in PR** #47246  

### To Reproduce

1. Set up impersonation - more or less as described [here](https://www.metabase.com/learn/metabase-basics/administration/permissions/impersonation#:~:text=Impersonation%20is%20a%20permissions%20setting,your%20database%20executes%20the%20query)
2. Obtain the database role user attribute from SSO/Keycloak/JWT JSONs attributes like {"groups": ["myFirstGroup", "mySecondGroup"]}
3. See error
```
class clojure.lang.PersistentVector cannot be cast to class java.lang.CharSequence (clojure.lang.PersistentVector is in unnamed module of loader 'app'; java.lang.CharSequence is in module java.base of loader 'bootstrap')
...
metabase_enterprise.advanced_permissions.driver.impersonation$connection_impersonation_role.invokeStatic(impersonation.clj:73)
```
This means that [this line of Metabase Enterprise code crashes](https://github.com/metabase/metabase/blob/6bf7bfca9aa8a71ba8e155a2c90713d0d9999951/enterprise/backend/src/metabase_enterprise/advanced_permissions/driver/impersonation.clj#L73).

### Expected behavior

**Expected result** would be at least a reasonable error, which data(type) is expected and at which point exactly input data is considered to be bad.

**An even nicer and customer friendly solution** would be to check the type of the role variable, and convert it as necessary. 

### Logs

```
class clojure.lang.PersistentVector cannot be cast to class java.lang.CharSequence (clojure.lang.PersistentVector is in unnamed module of loader 'app'; java.lang.CharSequence is in module java.base of loader 'bootstrap')
...
metabase_enterprise.advanced_permissions.driver.impersonation$connection_impersonation_role.invokeStatic(impersonation.clj:73)
```

### Information about your Metabase installation

```JSON
- Metabase version v1.50.21
- SSO via JWT/Keycloak/OIDC

Rather less relevant information:
- Metabase Hosting-Environment: OpenShift
- Metabase internal Database: MySQL
- Tested with metabase-clickhouse-driver
```

### Severity

blocking

**Comment by paoliniluis** [comment]
Hi there, we should be reviewing and shipping the fix soon, but I wanted to check the use case here as I'm not fully getting why you would like to pass 2 roles to a database. What's the use case here?

**Comment by paoliniluis** [comment]
@s-huk you're using clickhouse as the DW is that correct?

**Comment by s-huk** [comment]
@paoliniluis yes, and we have already taken a look at the [metabase-clickhouse-driver](https://github.com/ClickHouse/metabase-clickhouse-driver/blob/master/src/metabase/driver/clickhouse.clj#L274-L299)

**Comment by noahmoss** [comment]
Relabeled as feature request after discussion with product.

**Comment by s-huk** [comment]
Once again, I suggested a sound and very easy solution in PR #47246  

...as I already pointed out above: **An even nicer and customer friendly solution** would be to check the type of the role variable, and convert it as necessary.

**Comment by noahmoss** [comment]
@s-huk As I noted previously, concatenation of strings using commas is a driver implementation detail. Even _if_ every driver does this in the same way, they may have different rules around quoting individual roles, or other syntax differences. For instance, our Postgres driver already will [quote a role](https://github.com/metabase/metabase/blob/master/src/metabase/driver/postgres.clj#L968-L969) if it includes non-alphanumeric characters. 

It's plausible that a future driver could have both requirements—multiple roles and special quoting rules—in which case your proposed implementation would not work. Or a future might have a different syntax for multiple roles altogether.

I'm open to the idea of updating `driver/set-role!` and `driver/set-role-statement!` to take a list of strings in place of a single string. If you'd like to submit a new PR with this approach (or another approach which doesn't have the issue above) I'm happy to review it.

**Comment by s-huk** [comment]
I will consult with my team to determine whether to participate here, comparing it to other possible solutions.

It sounds like you want to change the type, i.e., first overload the set-role-statement method somehow and then gradually phase out the non-set method from every possible driver.

**Comment by s-huk** [comment]
@noahmoss
Before I contribute vainly to unnecessarily impairing the simplicity and clarity of the driver interface, I need to ask you again: **Are you really sure you want to change the type of the role variable from a string to a collection of strings?**

To reconsider this more clearly, let's focus on the following role input scenarios which can occur:

1. The role attribute is a string consisting of exactly one role.
2. The role attribute is a string consisting of multiple roles, comma-separated by the user.
3. The role attribute is a collection.

Instead of changing/extending the type of your interface's `role` variable, I would suggest simplifying scenario 3 into scenario 2.

As you point out, Metabase can't validate or transform input from scenario 2. Only the drivers can transform such strings according to their underlying database syntax, as you already mentioned. And why wouldn’t the drivers be able to handle this with, e.g., regular expressions instead of collections?

Therefore, I still recommend merging PR #47246 (which already maps scenario 3 into 2). Upon closer inspection, it still seems to be the most straightforward and less complicated solution. It works with all possible drivers as long as there is no more than one role involved (role input scenarios 1-3 would all be supported). It also already supports multiple roles in drivers like Clickhouse. As for Postgres, this system doesn't seem to support multiple roles anyway.

If there is a database with special quoting rules that supports multiple roles, we would simply need to adapt its driver to handle comma-separated strings - so what? (because this would have to be done anyways, to be able to handle input scenario 2)

**Comment by s-huk** [comment]
@dpsutton Hi, I didnt suggest "letting a string randomly be a collection" (just @noahmoss seems to think in this direction). PR #47246 solely proposes to homogenize it to a string if it accidentally turns out to be a collection. So sticking with your example, I would rather suggest to transform something like `{:email "sandbox@metabase.com", :id 2, :login_attributes {"id" ["1" "2"]}}` to `{:email "sandbox@metabase.com", :id 2, :login_attributes {"id" "1,2"}}`.

**Comment by Gerrit-K** [comment]
Chiming in from a different company, but with a very similar setup as @s-huk: we use Metabase Enterprise behind Keycloak to query data from ClickHouse. This bug (/ missing feature, if you like) is currently blocking our rollout of Connection Impersonation, which was one of the key features for us to buy the Enterprise license in the first place.

It's a pity that the existence of multiple roles was not considered for the first concept of this feature, because I think it's a rather common case. Users often have not just one distinct but many different, potentially even overlapping roles set via their IdP and all the DBMS I've come across support users with multiple roles, so IMO Metabase should support that, too.

Regarding the history of this ticket and the associated discussions, I can understand the arguments on both sides. I agree that joining the roles by a comma should be the responsibility of the specific driver instead of Metabase itself, but I also don't see that this (temporary?) bugfix would significantly impair functionality or pose a risk. The method was designed with one input role and one output string in mind, so it was never expected to work with a vector of roles anyways. Thus, I would not expect that changing the (unintended) vector output back to a string output breaks anything. Of course, this should be properly documented and tracked with an according cleanup task once the design plan is revised for multi-role scenarios. 

But for now, we would really appreciate a short-term fix to unblock our transition to Metabase Enterprise w/ Connection Impersonation. Regarding the workaround as recommended on [this PR comment](https://github.com/metabase/metabase/pull/47246#issuecomment-2322037240):
> Until we support this use case officially, you may be able to work around this by storing the concatenated list of roles as a string directly in the user attributes.

To my knowledge, this is not easily doable in Keycloak. It would require us to implement a custom User Attribute Mapper or a Script Mapper because Keycloak does not provide this functionality out of the box with its standard mappers.

**Comment by paoliniluis** [comment]
@Gerrit-K please contact our helpdesk

**Comment by s-huk** [comment]
@Gerrit-K Thank you for pointing this out (exactly my thoughts too):
> The method was designed with one input role and one output string in mind, so it was never expected to work with a vector of roles anyways. Thus, I would not expect that changing the (unintended) vector output back to a string output breaks anything. Of course, this should be properly documented and tracked with an according cleanup task once the design plan is revised for multi-role scenarios.

@dpsutton @noahmoss 
Sounds like a good plan to proceed, doesnt it?

And, yes, this is exactly what we have to deal with too:
> To my knowledge, this is not easily doable in Keycloak. It would require us to implement a custom User Attribute Mapper or a Script Mapper because Keycloak does not provide this functionality out of the box with its standard mappers.

**Comment by s-diehl** [comment]
@paoliniluis any updates on this?
